service/controller/clusterapi/v17/resources/machinedeploymentstatus/create.go

```diff
@@ -2,8 +2,103 @@ package machinedeploymentstatus
 
 import (
 	"context"
+
+	"github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	"github.com/giantswarm/tenantcluster"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/cluster-operator/service/controller/clusterapi/v17/controllercontext"
+	"github.com/giantswarm/cluster-operator/service/controller/clusterapi/v17/key"
 )
 
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToMachineDeployment(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var nodes []corev1.Node
+	var ready []corev1.Node
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding nodes of tenant cluster")
+
+		l, err := cc.Client.TenantCluster.K8s.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenantcluster.IsTimeout(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "timeout fetching certificates")
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+
+			return nil
+		} else if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant API not available")
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		nodes = l.Items
+
+		for _, n := range nodes {
+			for _, c := range n.Status.Conditions {
+				if c.Type == corev1.NodeReady && c.Status == corev1.ConditionTrue {
+					ready = append(ready, n)
+				}
+			}
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found nodes of tenant cluster")
+
+		replicasChanged := cr.Status.Replicas != int32(len(nodes))
+		readyReplicasChanged := cr.Status.ReadyReplicas != int32(len(ready))
+
+		if !replicasChanged && !readyReplicasChanged {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "not updating status of machine deployment")
+			return nil
+		}
+	}
+
+	var resourceVersion string
+	{
+		m, err := r.cmaClient.ClusterV1alpha1().MachineDeployments(cr.Namespace).Get(cr.Name, metav1.GetOptions{})
```

Comment:
Just to double check: `cluster-operator` is the only component that updates status of `MachineDeployment`? If not, I think refreshing `resourceVersion` here could lead to race condition.

Reply:
The idea is that everything non-provider specific is managed by `cluster-operator`. The `MachineDeployment` status as well as the generic `Cluster` status should be managed here. 

Then, the resource version update is the usual pattern used to update CR statuses. In case we would not use the latest version for updates, the update would fail. I do not follow how this should overwrite anything. We do this basically everywhere and the resource version is the primitive to "lock" mutations.

Reply:
I don't think we earlier updated same resource from different programs. What I had in my mind was something like this:

1. `cluster-operator` starts reconciliation with `resourceVersion:7`
2. `aws-operator` updates status and version becomes `resourceVersion:8`
3. `cluster-operator` holds old CR object but updates its version to `resourceVersion:8`
4. Now the CR has `resourceVersion:9` but contents are `resourceVersion:7` + changes made by `cluster-operator`.

This happens because this function holds CR object from reconciliation execution in `cr` and only replaces its `resourceVersion` with the latest one. This specifically breaks the premise by optimistic locking which would work correctly when the `resourceVersion` is not updated.

If the `resourceVersion` is not refreshed and `aws-operator` updates the CR in-between as proposed above, then line 92 below would fail because that version doesn't match the one in present CR.

Reply:
When RV is 9 and you update using 7 you get an error. We have these situations in resources managing k8s native runtime objects like services. The resources fail, get retried and succeed eventually. This is how it works everywhere right now. The only reason why anything could go wrong with this would be when different processes manage the very same properties. That would be broken by design. We do not do this. The `machinedeploymentstatus` resource manages the `MachineDeployment` `Replicas` and `ReadyReplicas` as of this proposal. No other resource does it and should do it.

Reply:
How does `Update()` know that above code didn't modify unrelated field here when the `resourceVersion` was changed?

Current implementation would work in desired way if it was sending individual fields as patches. Due to refreshing of `resourceVersion` it would be still wrong from pure correctness point of view, but in our use case it would work in desired way (i.e. last write wins).

The problem here is that this code calls `Update()` below which gets whole object as parameter and if fields that `cluster-operator` doesn't manage were changed during the execution here, then those fields get reverted after the `Update()` call because neither `client-go` implementation nor k8s `apiserver` can know that this code didn't touch those fields here (because `revisionVersion` was refreshed and that was the only gatekeeper for correctness).

> When RV is 9 and you update using 7 you get an error.

And that specifically is the way how optimistic locking works. That's the only way to maintain correctness.

> The resources fail, get retried and succeed eventually.

And when on retry the code reads latest version of object, only changes the fields relevant and retries the `Update()`, then it works correctly.

> This is how it works everywhere right now. 

If we do this exactly like here, then we have it wrong everywhere and we have just been lucky because we didn't try to modify same object from different programs (managing different fields) at the same time.

Reply:
We chatted and I got the concerns now. We paired a bit and fixed the issue.

Reply:
We chatted and I got the concerns now. We paired a bit and fixed the issue.